pkg/querier/remote_read.go

```diff
@@ -147,14 +147,72 @@ func remoteReadStreamedXORChunks(
 	}
 	w.Header().Set("Content-Type", api.ContentTypeRemoteReadStreamedChunks)
 
+	// Process all queries concurrently and collect their ChunkSeriesSet results
+	type queryResult struct {
+		idx    int
+		series storage.ChunkSeriesSet
+		err    error
+	}
+
+	resultCh := make(chan queryResult, len(req.Queries))
+
+	// Launch goroutines to process each query concurrently
 	for i, qr := range req.Queries {
-		if err := processReadStreamedQueryRequest(ctx, i, qr, q, w, f, maxBytesInFrame); err != nil {
+		go func(i int, qr *prompb.Query) {
+			start, end, _, _, matchers, hints, err := queryFromRemoteReadQuery(qr)
+			if err != nil {
+				resultCh <- queryResult{idx: i, err: err}
+				return
+			}
+
+			querier, err := q.ChunkQuerier(int64(start), int64(end))
+			if err != nil {
+				resultCh <- queryResult{idx: i, err: err}
+				return
+			}
+
+			// The streaming API has to provide the series sorted.
+			seriesSet := querier.Select(ctx, true, hints, matchers...)
+			resultCh <- queryResult{idx: i, series: seriesSet, err: nil}
+		}(i, qr)
+	}
+
+	// Collect results and maintain order
+	results := make([]queryResult, len(req.Queries))
+	var lastErr error
+
+	for range req.Queries {
+		result := <-resultCh
+		results[result.idx] = result
+		if result.err != nil {
+			lastErr = result.err
+		}
+	}
+
+	// If any query failed, return the error
+	if lastErr != nil {
+		code := remoteReadErrorStatusCode(lastErr)
+		if code/100 != 4 {
+			level.Error(logger).Log("msg", "error while processing remote read request", "err", lastErr)
+		}
+		http.Error(w, lastErr.Error(), code)
+		return
+	}
+
+	// Stream the results in order
+	for i, result := range results {
+		if err := streamChunkedReadResponses(
+			prom_remote.NewChunkedWriter(w, f),
+			result.series,
+			i,
+			maxBytesInFrame,
+		); err != nil {
 			code := remoteReadErrorStatusCode(err)
 			if code/100 != 4 {
-				level.Error(logger).Log("msg", "error while processing remote read request", "err", err)
+				level.Error(logger).Log("msg", "error while streaming remote read response", "err", err)
 			}
 			http.Error(w, err.Error(), code)
-			return
+			break
 		}
 	}
 	w.WriteHeader(http.StatusOK)
```

Comment:
Is this call effective if we've already written some data to `w` in `streamChunkedReadResponses` above?

Same for the `http.Error` call above.

Reply:
this was a leftover from another set of changes. Restored to `return` now

Reply:
oh i didn't fully comprehend what you were saying, sorry. yes, you're right. pushed a change. we can't really do much about this, but at least I added comments and put the status code in an earlier place. See https://github.com/grafana/mimir/pull/11732/commits/e9495d66cbef7b787a7fe2fcf0372238eba832e0

Reply:
actually, i removed the status setting altogether, because the tests I had asserted that an error before we start streaming will be reflected in the status code. `http.ResponseWriter.Write()` will set the status code to 200 on the first `Write()` if it's not set already

this is the new commit https://github.com/grafana/mimir/pull/11732/commits/84664c4b9486078dcfae1bc0277e52c123cf835b